crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -312,34 +328,108 @@ pub(super) fn infer_same_file_expression_type<'db>(
 ///
 /// Use [`infer_same_file_expression_type`] if it is guaranteed that  `expression` is in the same
 /// to avoid unnecessary salsa ingredients. This is normally the case inside the `TypeInferenceBuilder`.
-#[salsa::tracked(cycle_fn=single_expression_cycle_recover, cycle_initial=single_expression_cycle_initial, heap_size=ruff_memory_usage::heap_size)]
 pub(crate) fn infer_expression_type<'db>(
     db: &'db dyn Db,
     expression: Expression<'db>,
+    tcx: TypeContext<'db>,
 ) -> Type<'db> {
-    let file = expression.file(db);
+    infer_expression_type_impl(db, InferExpression::new(db, expression, tcx))
+}
+
+#[salsa::tracked(cycle_fn=single_expression_cycle_recover, cycle_initial=single_expression_cycle_initial, heap_size=ruff_memory_usage::heap_size)]
+fn infer_expression_type_impl<'db>(db: &'db dyn Db, input: InferExpression<'db>) -> Type<'db> {
+    let file = input.expression(db).file(db);
     let module = parsed_module(db, file).load(db);
 
     // It's okay to call the "same file" version here because we're inside a salsa query.
-    infer_same_file_expression_type(db, expression, &module)
+    let inference = infer_expression_types_impl(db, input);
+    inference.expression_type(input.expression(db).node_ref(db, &module))
 }
 
 fn single_expression_cycle_recover<'db>(
     _db: &'db dyn Db,
     _value: &Type<'db>,
     _count: u32,
-    _expression: Expression<'db>,
+    _input: InferExpression<'db>,
 ) -> salsa::CycleRecoveryAction<Type<'db>> {
     salsa::CycleRecoveryAction::Iterate
 }
 
 fn single_expression_cycle_initial<'db>(
     _db: &'db dyn Db,
-    _expression: Expression<'db>,
+    _input: InferExpression<'db>,
 ) -> Type<'db> {
     Type::Never
 }
 
+/// An `Expression` with an optional `TypeContext`.
+///
+/// This is a Salsa supertype used as the input to `infer_expression_types` to avoid
+/// interning an `ExpressionWithContext` unnecessarily when no type context is provided.
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq, salsa::Supertype, salsa::Update)]
+enum InferExpression<'db> {
```

Comment:
We also talked about potentially splitting the queries, but this supertype is effectively a more convenient way of doing that (internally, Salsa essentially creates a query for each variant of the input enum). The performance impact of this PR seems to be very minor with this optimization (and without it, there was a noticeable %2-3 regression).

Reply:
This is a nice solution!